File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        
        
        // using the tortoise and hare concept
        
        ListNode* slow  = head, *fast = head;
        
        
        if(head == NULL || head->next == NULL){
            // edge case
            return NULL;
        }
        
        // traverse and find the meeting point
        
        // to terminate the loop, the fast should not be null
        
        while(fast->next != NULL && fast->next->next != NULL){
            
            fast = fast->next->next;
            slow = slow->next;
            
            if(slow == fast){
                break;
            }
            
        }
        
        // now the loop has been terminated
        
        if(fast->next == NULL || fast->next->next == NULL){   
            // then there is no loop
            return NULL;
        }
        
        
        // loop is there
        
        fast = head;
        
        
        while(slow != fast){
            slow = slow->next;
            fast = fast->next;
        }
        
        
        return slow;
    }
};

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow = head
        fast = head

        # if slow.next == None or fast.next.next == None: 
            # single element or 2 elements only
            # return None

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

            if slow == fast:
                break
        else:
            return None    

        # now put one at the start and start another loop
        slow = head

        while slow != fast:
            slow = slow.next
            fast = fast.next

        return fast

# Synced seamlessly with LeetHub Pro
# Pro features: https://bit.ly/leethubpro | Free version: https://bit.ly/leethubv4
# Get it here: https://chromewebstore.google.com/detail/bcilpkkbokcopmabingnndookdogmbna